REPL: upgrade to JLine 3 (benefits include multi-line editing)

som-snytt opened the pull request:
JLine 3 supports multi-line editing, a better tab-completion UI, and more.

Configure keybindings with `-Xjline:emacs` (the default) or `-Xjline:vi`; disable with `-Xjline:off`

History file is now `~/.scala_history_jline3`

`scala-compiler.jar` now depends on `net.java.dev.jna#jna` (instead of `org.fusesource.jansi#jansi`, as formerly).

Minimal changes to support new API, with basic completion and history.

Fixes scala/bug#11367. Rebased from https://github.com/scala/scala/pull/7645

som-snytt commented:
/rebuild

SethTisue commented:
I salivate every time I run across this in the PR queue.

adriaanm commented:
Took care of one tiny little paper cut in the macro-annot tests. Sorry, I didn't mean to add myself as the committer for just a one line change, but don't know how to convey this to git.

som-snytt commented:
I actually have a delta to push for this.

adriaanm commented:
Ok, I will hold off now that I have a ✅  
Hope my push -f didn't cause any conflicts

adriaanm commented:
Actually, one more commit for the Reader header.

som-snytt commented:
Actually, that last force was it, so have at it. I have a sense that there are several things to do, but I didn't make a checklist.

adriaanm commented:
> I didn't make a checklist.

I believe in fact you did, though in a different lifetime, as it were.

adriaanm commented:
I should add that I've been trying the REPL from this PR -- looks ready to ship in 2.13.1 to me! Really, really nice work. Multi-line history ftw!

som-snytt commented:
Thanks, @adriaanm. I really ran out of steam on this, sorry; maybe because not knowing if it would benefit anyone. There are a few conflicting signals about what's envisioned for either dotty or ammonite. Personally and professionally, I totes do use REPL as an internal tooling interface, i.e., as a command line tool. I'm willing to contribute further baby steps.

Speaking of ✅ today's nytimes crossword has answers like "checks and balances" where the joke is that "check" is a single square. The "revealer" is "check all the boxes". I noted in the comments that it would be nice if the crossword supported Unicode. It would be nice if all our tools supported Unicode.

adriaanm commented:
We are pretty committed to the current repl for the 2 series. It’s really
great to be on jline 3! I was a little too eager with my parse error
suppression in the last commit (there was an issue with lines with a
leading dot). It can probably just be dropped and refined later. I’m
currently on 👶leave, but the plan is to implement a few of the ideas
pioneered by ammonite in our own repl this summer.

On Thu, 27 Jun 2019 at 22:15, som-snytt <notifications@github.com> wrote:

> Thanks, @adriaanm <https://github.com/adriaanm>. I really ran out of
> steam on this, sorry; maybe because not knowing if it would benefit anyone.
> There are a few conflicting signals about what's envisioned for either
> dotty or ammonite. Personally and professionally, I totes do use REPL as an
> internal tooling interface, i.e., as a command line tool. I'm willing to
> contribute further baby steps.
>
> Speaking of ✅ today's nytimes crossword has answers like "checks and
> balances" where the joke is that "check" is a single square. The "revealer"
> is "check all the boxes". I noted in the comments that it would be nice if
> the crossword supported Unicode. It would be nice if all our tools
> supported Unicode.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/scala/scala/pull/8036?email_source=notifications&email_token=AAAWHS6VVRJ2TW35TFEBQ3LP4UNXNA5CNFSM4HK5YVZ2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYYHZSY#issuecomment-506494155>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAWHSZG4FPMD4PBHXNYYIDP4UNXNANCNFSM4HK5YVZQ>
> .
>

lrytz commented on build.sbt:
```diff
@@ -44,7 +44,9 @@ val junitDep          = "junit"                          % "junit"
-val jlineDep          = "jline"                          % "jline"                            % versionProps("jline.version")
+val jlineDep          = "org.jline"                      % "jline"                            % versionProps("jline.version")
+val jansiDep          = "org.fusesource.jansi"           % "jansi"                            % versionProps("jansi.version")
+val jnaDep            = "net.java.dev.jna"               % "jna"                              % versionProps("jna.version")
```
https://github.com/jline/jline3#jansi-vs-jna says

>  If one of the Jansi or JNA library is present, it will be used and JLine will use native calls

> There is no difference between JLine's support for Jansi and JNA. Both will provide the exact same behaviors. So it's a matter of preference

So do we need both?

som-snytt replied:
I remember I tried the simple thing and it didn't work, but that was a long time ago. I'll try to refresh some memory cells and also experiment.

adriaanm replied:
It looks like coursier uses [just jansi](https://github.com/coursier/coursier/commit/e9d81d81130f0280adb39525e4b9dba7f9a3bd06) -- maybe this can inspire?

SethTisue replied:
spun out to https://github.com/scala/scala-dev/issues/677

lrytz reviewed (approved):
I think we should just get this in so that it gets tested by those working on 2.13.x. Looking through the change at a high level, i only have one question, about the dependencies.

som-snytt commented:
I guess Adriaan is in :man: :baby: :fog:, as in Karl the fog not as in head in the clouds, which also explains his heady enthusiasm, though we'd have to run some benchmarks to accurately compare to his previous levels.

I'll check out the previous commit and achieve minimal merge polish, with confidence that only @dwijnand is likely to complain about breakage.

dwijnand commented:
What breakage am I likely to complain about?  The only one that comes to mind is source-breakage of zinc's "console" abstraction of the REPL, which we'd have to address.

Other than that I don't have any PRs touching the same files (merge conflict breakage) or PRs looking to make changes to any behaviour here (semantics breakage).  Did you mean the Zinc thing or am I forgetting something?

som-snytt commented:
Viewing my previous comment in Windows, I don't especially like 🌫 but maybe I'd like it more as a moustache.

Sorry @dwijnand let me add that adding your handle in FIrefox on Windows was painful, which was the whole reason we fought the browser wars.

Doubly sorry, now I don't remember what I was responding to. Please take my word that it was humorous to at least one of us, and that if I see a real issue in future I will report it.

adriaanm commented:
@SethTisue AFAICT the last action item before this can go into 2.13.1 is to see if we can restrict our additional dependencies to just [jansi](https://github.com/scala/scala/pull/8036#discussion_r306728108), and drop jna. 

Even with my brain 🌩 starting to clear, I still think we should get this in 😁

som-snytt commented on build.sbt:
```diff
@@ -173,6 +175,7 @@ lazy val commonSettings = instanceSettings ++ clearSourceAndResourceDirectories
     "-doc-source-url", s"https://github.com/scala/scala/tree/${versionProperties.value.githubTree}€{FILE_PATH_EXT}#L€{FILE_LINE}"
   ),
   //maxErrors := 10,
+  maxErrors := 10,
```
Cleanup.

som-snytt commented on src/repl-frontend/scala/tools/nsc/interpreter/jline/Reader.scala:
```diff
@@ -0,0 +1,222 @@
+    }
+    try history.attach(reader)
+    catch {
+      case e: IllegalArgumentException if e.getMessage.contains("Bad history file syntax") =>
```
This is logged now instead of thrown.

eed3si9n commented on src/repl-frontend/scala/tools/nsc/interpreter/shell/ShellConfig.scala:
```diff
@@ -55,13 +57,16 @@ trait ShellConfig {
   // This property is used in TypeDebugging. Let's recycle it.
   val colorOk = Properties.coloredOutputEnabled
 
+  val historyFile = s"$userHome/.scala_history"
```
What do you think about picking a different name for this file, so we can keep using the old file name for older Scala version's REPL.

som-snytt replied:
We can disable history timestamps for jline and share the (jline default) history file. There is no `-D` for history file name; 2.12 FileBackedHistory had `-Dscala.shell.histfile=`.

som-snytt replied:
Yes, i.e., this was a quick hack. I don't know if there is a use case for timestamped REPL lines, but what people would want is versioned history. "I've been futzing with these type parameters for an hour, show me all the things I've tried so far." Or show me the versions that compiled and produced a result.

SethTisue replied:
spun out as https://github.com/scala/scala-dev/issues/682

retronym commented on src/repl/scala/tools/nsc/interpreter/PresentationCompilation.scala:
```diff
@@ -89,7 +89,7 @@ trait PresentationCompilation { self: IMain =>
   private var lastCommonPrefixCompletion: Option[String] = None
 
-  abstract class PresentationCompileResult(val compiler: interactive.Global, val inputRange: Position, val cursor: Int, val buf: String) extends PresentationCompilationResult {
+  private abstract class PresentationCompileResult(val compiler: interactive.Global, val inputRange: Position, val cursor: Int, val buf: String) extends PresentationCompilationResult {
```
I intended this to be usable by tools build on top of the REPL (web UIs for instance) that need to be tightly coupled to compiler internals but want the convenience of this API. So I'd suggest to leave it public.

som-snytt replied:
OK thanks. I assume I found it confusing because of the interface. All the names will be ironed out in the REPL protocol.

retronym commented on src/repl-frontend/scala/tools/nsc/interpreter/jline/Reader.scala:
```diff
@@ -0,0 +1,223 @@
+      .completer(completer)
+      .history(history)
+      .parser(parser)
+      .terminal(terminal)
```
In our usage of JLine2, we picked up configuration from `~/.inputrc` [etc](https://github.com/jline/jline2/blob/12b98d94589e3bd6a60719c495b0ab911c571329/src/main/java/jline/console/ConsoleReader.java#L290-L306).

JLine3 does have support for parsing this file since https://github.com/jline/jline3/commit/9ccfe0b121d2ea72a8524c84a9bda5abaa68d45a, but it looks like we need to [call `InputRC.configure`](https://github.com/jline/jline3/commit/9ccfe0b121d2ea72a8524c84a9bda5abaa68d45a#diff-b26d899b1659f142e858937634634a4bR56) after determining which config file to use ourselves.

Furthermore, in the current configuration file, sections guarded with `$if JLine .... $endif` are interpreted by the Scala REPL. Under this change, config would need to be moved to `$if scala` (the `appname` used here. That's probably fine if release noted.

Here's my `.inputrc` for reference.

```
$if Bash
	"\e[A": history-search-backward
	"\e[B": history-search-forward
    Space: magic-space
$endif

#set blink-matching-paren on
"\C-xq": "\eb\"\ef\""

"\C-x)": "\ea\(\ee\)"

$if JLine
	"\C-xu": "import scala.tools.reflect.ToolBox; val m = scala.reflect.runtime.currentMirror; val u = m.universe; import u._; val tb = m.mkToolBox()"
	"\C-x)": "\C-a(\C-e)"
	"\C-x}": "\C-a{\C-e}"
	"\C-x{": "{}\C-b"
	"\C-x(": "()\C-b"

$endif
```

som-snytt replied:
Thanks. I know there are tickets that turn on local `.inputrc` but this didn't come up.

SethTisue replied:
spun out as https://github.com/scala/scala-dev/issues/683

retronym commented:
Previously: `println<TAB><TAB>` would show the definition string for the methods with names that exactly matched.

In the JLine3 repl, this suggestion shows up after a single `<TAB>`, and additional `<TAB>` lets cycles through those options with ZSH-like menu completion. Selecting one puts the def string into the buffer which isn't what we're after.

It was a hack to display this help through JLine2's completion system so we should rethink our UI options here with JLine3.

retronym commented:
Completion suggestions have changed when the cursor is midway through an identfier.

Before:

```
scala> 1.toD<TAB>XXXX
toDegrees   toDouble

scala> 1.toD<TAB>ouble
toDegrees   toDouble
```

This PR:

```
scala> 1.toD<TAB>XXXX. // no completions

scala> 1.toD<TAB>ouble. // "ouble" is completed
```
```

retronym commented on src/repl-frontend/scala/tools/nsc/interpreter/jline/Reader.scala:
```diff
@@ -0,0 +1,223 @@
+    def backupHistory(): Unit = {
+      import java.nio.file.{Files, Paths, StandardCopyOption}, StandardCopyOption.REPLACE_EXISTING
+      val hf = Paths.get(config.historyFile)
+      val bk = Paths.get(config.historyFile + ".bk")
```
When creating the history file, we [change its permissions](https://github.com/scala/scala/blob/515916c42f8f2f5c476ffabd6774e5a903feda10/src/repl-jline/scala/tools/nsc/interpreter/jline/FileBackedHistory.scala#L32-L44) to `600`. This is a security measure for multi-user systems that don't secure the `$HOME` by default (I'm looking at you OS/X and Windows).

 We should do the same with the backup and the new copy of the history file.

som-snytt replied:
Yes, this was a quick hack.

SethTisue replied:
spun out as https://github.com/scala/scala-dev/issues/684

szeiger commented:
This doesn't work for me on Windows under MSYS:

```
$ ./build/quick/bin/scala -nc
Welcome to Scala 2.13.1-20190826-185208-4346a44 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_121).
Type in expressions for evaluation. Or try :help.
Aug 27, 2019 4:58:38 PM org.jline.utils.Log logr
WARNING: Unable to retrieve infocmp for type msys
java.io.IOException: Cannot run program "infocmp.exe": CreateProcess error=2, The system cannot find the file specified
        at java.lang.ProcessBuilder.start(Unknown Source)
        at org.jline.utils.InfoCmp.getInfoCmp(InfoCmp.java:547)
        at org.jline.terminal.impl.AbstractTerminal.parseInfoCmp(AbstractTerminal.java:187)
        at org.jline.terminal.impl.AbstractWindowsTerminal.<init>(AbstractWindowsTerminal.java:91)
        at org.jline.terminal.impl.jansi.win.JansiWinSysTerminal.<init>(JansiWinSysTerminal.java:76)
        at org.jline.terminal.impl.jansi.win.JansiWinSysTerminal.createTerminal(JansiWinSysTerminal.java:67)
        at org.jline.terminal.impl.jansi.JansiSupportImpl.winSysTerminal(JansiSupportImpl.java:129)
        at org.jline.terminal.TerminalBuilder.doBuild(TerminalBuilder.java:345)
        at org.jline.terminal.TerminalBuilder.build(TerminalBuilder.java:259)
        at org.jline.terminal.TerminalBuilder.terminal(TerminalBuilder.java:75)
        at scala.tools.nsc.interpreter.jline.Reader$.apply(Reader.scala:58)
        at scala.tools.nsc.interpreter.shell.ILoop.defaultIn$lzycompute(ILoop.scala:71)
        at scala.tools.nsc.interpreter.shell.ILoop.defaultIn(ILoop.scala:66)
        at scala.tools.nsc.interpreter.shell.ILoop.run(ILoop.scala:953)
        at scala.tools.nsc.MainGenericRunner.runTarget$1(MainGenericRunner.scala:87)
        at scala.tools.nsc.MainGenericRunner.run$1(MainGenericRunner.scala:91)
        at scala.tools.nsc.MainGenericRunner.process(MainGenericRunner.scala:103)
        at scala.tools.nsc.MainGenericRunner$.main(MainGenericRunner.scala:108)
        at scala.tools.nsc.MainGenericRunner.main(MainGenericRunner.scala)
Caused by: java.io.IOException: CreateProcess error=2, The system cannot find the file specified
        at java.lang.ProcessImpl.create(Native Method)
        at java.lang.ProcessImpl.<init>(Unknown Source)
        at java.lang.ProcessImpl.start(Unknown Source)
        ... 19 more
```

retronym commented:
Another change in completion behaviour:

```
scala> "".getClass.getDM<TAB>
```

Used to expand to:

```
scala> "".getClass.getDeclaredMethod
getDeclaredMethod   getDeclaredMethods
```

But now offers no completions. AFAICT JLine is [filtering candidates on our behalf.](https://github.com/jline/jline3/blob/master/reader/src/main/java/org/jline/reader/impl/LineReaderImpl.java#L4184-L4230) without an obvious extension point to add our own strategy.

som-snytt commented:
TIL on 2.12 `scala> java.util.concurrent.Execututors.` unexpectedly offers completions for the misspelled member but repl doesn't coordinate to correct the spelling.

lrytz commented:
It would be really nice to have this in 2.13.2! (4-5 weeks)

som-snytt commented:
I looked at this recently, trying to close out PRs, and it would have been nice in 2.13.1 but is depressing to revisit again. I need it so that companions entered on different lines drops you into a multiline edit. But otherwise, the creaky old REPL won't move any needles.

SethTisue commented:
the next step here is to merge or rebase with #8600 — that's where the merge conflicts start.

As a new baseline, I've pushed merges that bring it to immediately before that point. (And I checked that the REPL still functions.)

som-snytt commented:
Oh man, why do people keep messing with REPL...

SethTisue commented:
@som-snytt I'm just getting started with helping with this. more to come. I'm tasked with this now, it's no longer on an if-I-ever-get-around-to-it basis

som-snytt commented:
I was just joking about how my old PRs keep stomping on my new PRs, clarifying for @dwijnand . Or maybe it's the other way around.

I'm not sure how best to help. I can spool up again.

It's like when Arrow says "Suit up!" except it's, "Spool up!"

som-snytt commented:
I only hope they won't ask me to test it on windoze.

martijnhoekstra commented:
> I only hope they won't ask me to test it on windoze.

I'm happy to test (and try to fix if broken) on windoze

SethTisue commented:
rebased. two test failures, one trivial, the other looks like I failed to merge #8621 correctly. digging...

SethTisue commented:
yay, tests passing.

I've read through all the review comments and I'm pulling together a list of remaining work items. looks like tab completion is going to need some work.

martijnhoekstra commented:
Windows, both MSYS and cmd:

```
$ ./build/quick/bin/scala
Welcome to Scala 2.13.2-20200226-194149-2ef554a (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_201).
Type in expressions for evaluation. Or try :help.
Exception in thread "main" java.io.UncheckedIOException: java.nio.charset.MalformedInputException: Input length = 1
        at java.io.BufferedReader$1.hasNext(BufferedReader.java:574)
        at java.util.Iterator.forEachRemaining(Iterator.java:115)
        at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
        at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:580)
        at org.jline.reader.impl.history.DefaultHistory.load(DefaultHistory.java:86)
        at org.jline.reader.impl.history.DefaultHistory.attach(DefaultHistory.java:69)
        at scala.tools.nsc.interpreter.jline.Reader$.apply(Reader.scala:95)
        at scala.tools.nsc.interpreter.shell.ILoop.defaultIn$lzycompute(ILoop.scala:67)
        at scala.tools.nsc.interpreter.shell.ILoop.defaultIn(ILoop.scala:62)
        at scala.tools.nsc.interpreter.shell.ILoop.run(ILoop.scala:927)
        at scala.tools.nsc.MainGenericRunner.runTarget$1(MainGenericRunner.scala:87)
        at scala.tools.nsc.MainGenericRunner.run$1(MainGenericRunner.scala:91)
        at scala.tools.nsc.MainGenericRunner.process(MainGenericRunner.scala:103)
        at scala.tools.nsc.MainGenericRunner$.main(MainGenericRunner.scala:108)
        at scala.tools.nsc.MainGenericRunner.main(MainGenericRunner.scala)
Caused by: java.nio.charset.MalformedInputException: Input length = 1
        at java.nio.charset.CoderResult.throwException(CoderResult.java:281)
        at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:339)
        at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
        at java.io.InputStreamReader.read(InputStreamReader.java:184)
        at java.io.BufferedReader.fill(BufferedReader.java:161)
        at java.io.BufferedReader.readLine(BufferedReader.java:324)
        at java.io.BufferedReader.readLine(BufferedReader.java:389)
        at java.io.BufferedReader$1.hasNext(BufferedReader.java:571)
        ... 14 more

```

martijnhoekstra commented:
looks like my `~/.scala_history` is win1252 encoded. Anyone happen to know by heart if that's by default in platform encoding (and changed)?

martijnhoekstra commented:
both MSSYS and CMD: getting rid of history gives me

```
$ ./build/quick/bin/scala
Welcome to Scala 2.13.2-20200226-194149-2ef554a (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_201).
Type in expressions for evaluation. Or try :help.

```

No prompt, typing/enter doesn't show anything, ctrl-C exits the non-responsive scala shell and echoes what I typed there in the host shell.

SethTisue commented:
thanks, Martijn. I need to get my own Windows setup going (I have escaped this for the past 3 years or so, but it couldn't last forever) so I can help test. 

I think there are enough different issues here that need to be resolved before we can merge this, that I'm going to open a temporary repo, SethTisue/jline3-upgrade or something, to track them individually. but I'll wait and do it after GitHub is running smoothly again.

SethTisue commented:
> I'm going to open a temporary repo, SethTisue/jline3-upgrade or something, to track them individually

Lukas suggested using a label in scala-dev instead. I'll do that.

SethTisue commented:
from now on let's use https://github.com/scala/scala-dev/labels/t%3Ajline3 to discuss specifics

SethTisue commented:
on the label, we're using the "prio:low" label to indicate that something isn't a blocker for 2.13.2. we can triage again after we ship.

SethTisue committed:
use newer JLine version

the new features in 3.13 and 3.14 don't look especially
interesting offhand, but maybe there's something cool
I didn't spot, and we'll want all the bugfixes?

and update IntelliJ stuff for new JLine & Jansi versions

also update legal stuff.  JLine 3 is now BSD 3-clause not BSD.  and
bring License.rtf up to date with LICENSE.md.  (apparently the MSI
installer stuff needs the RTF version, sigh)

and it's not strictly JLine related but while I was updating legal
stuff I removed ScalaCheck since we currently neither in-source it nor
ship it

SethTisue committed:
JLine 3: history handling improvements

use a different filename than we used with JLine 2, since the format
is different (has timestamps) and because old files might have a
different character encoding, we encountered that in testing

also bring back code that restricts permissions on history file,
this went missing during the upgrade

hvesalai commented on src/repl-frontend/scala/tools/nsc/interpreter/shell/ILoop.scala:
```diff
@@ -183,11 +182,12 @@ class ILoop(config: ShellConfig, inOverride: BufferedReader = null,
+    cmd("completions", "<string>", "output completions for the given string", completionsCommand),
+    // TODO maybe just drop these commands, as jline subsumes them -- before reenabling, finish scala.tools.nsc.interpreter.jline.HistoryAdaptor
+    //cmd("edit", "<id>|<line>", "edit history", editCommand),
+    //historyCommand,
```
Current `emacs-sbt-mode` is integrated to the history command. It loads the history to emacs using the `:history` command and expects also commands executed through emacs to be there.

SethTisue replied:
@hvesalai thanks — please say so on https://github.com/scala/scala-dev/issues/690

hvesalai commented on src/repl-frontend/scala/tools/nsc/interpreter/shell/ILoop.scala:
```diff
@@ -529,26 +530,21 @@ class ILoop(config: ShellConfig, inOverride: BufferedReader = null,
-      echo(completionLines.mkString("\n"))
-    }
-
+    completions.candidates.map(c => s"[completions] $prefix$c").foreach(echo)
```
This needs to be reverted.

Echo flushes between lines and there is a huge difference whether you flush after each line or only once.

I remember writing the code first as you are now proposing and changing to what it is currently in master because of a good reason.

SethTisue replied:
som-snytt/scala@d1f4a72cf9821a22b6705ad46b4ad47c261c2062 LGTY?

hvesalai reviewed (changes requested):
The echo loop in completions command needs to be reverted.

SethTisue committed:
JLine 3: be emacs friendly (adjust :completions support)

based on review feedback from hvesalai

eed3si9n committed:
JLine 3: use JNA not Jansi

Goal: avoid conflicting with sbt's own Jansi version

Ref https://github.com/scala/scala-dev/issues/678

https://github.com/jline/jline3#jansi-vs-jna says

> On the Windows platform, relying on native calls is mandatory, so
> you need to have either Jansi or JNA library in your classpath along
> with the `jline-terminal-jansi` or `jline-terminal-jna` jar.

When I tested on Windows 7 or Windows 10 using `sbt console` I got
"WARNING: Unable to create a system terminal, creating a dumb terminal
(enable debug logging for more information)" and key input also didn't
work. (This could be more to do with sbt). I am guessing that Jansi
used by sbt itself maybe interfering with Jansi that gets loaded from
the compiler. Normally these duplicates are resolved by creating a
sandbox ClassLoader, but maybe it doesn't exactly work for native
libraries loaded into the same process.

additional changes by Seth:
* change Jansi -> JNA in legal info, Intellij config

Co-authored-by: Seth Tisue <seth@tisue.net>

eed3si9n commented:
I just sent a PR-to-PR to use JNA - https://github.com/som-snytt/scala/pull/5

SethTisue commented:
as per my remarks at https://github.com/scala/scala-dev/issues/677 , I've added Eugene's jna-not-jansi changes (squashed into one commit) to this PR. it seems like the likeliest way forward at present.

eed3si9n commented on src/repl-frontend/scala/tools/nsc/interpreter/jline/Reader.scala:
```diff
@@ -54,8 +54,7 @@ object Reader {
 
-    //val terminal = TerminalBuilder.builder().build()
-    val terminal  = TerminalBuilder.terminal() // defaults for now
+    val terminal = TerminalBuilder.builder().jna(true).build()
```
If JNA is optional, then this change I made might no longer be good?
https://github.com/jline/jline3/blob/master/terminal/src/main/java/org/jline/terminal/TerminalBuilder.java

SethTisue replied:
It's only intended to be completely optional for people who are using scala-compiler.jar without having any interest in the REPL at all. So I think it's okay.

som-snytt committed:
Use JLine3

Minimal changes to support new API,
with basic completion and history.
JLine handles continuation line edit.

SethTisue commented:
rebased and partially squashed. if all the commits pass CI, I think we could go ahead and merge this so that we'll all be testing it (and then address [remaining blockers](https://github.com/scala/scala-dev/issues?q=is%3Aopen+label%3At%3Ajline3+label%3Ablocker) in followup PRs)

SethTisue commented:
/rebuild 68931c3

SethTisue commented:
/rebuild 68931c3

SethTisue commented:
/synch

lrytz commented:
> * make REPL's JNA & JLine dependencies optional.  most users of scala-compiler.jar just want the compiler and aren't embedding the REPL, so there's no need to pull REPL stuff onto their classpath (and doing so might even cause a conflict. if someone is embedding the REPL and does want JNA, they can take care of adding the optional dependency themselves.

I think this is too breaking, at least for JLine, but IMO also for JNA. The dependency on JLine is not optional, as we get

```
$> java -cp `cs fetch -p -r https://scala-ci.typesafe.com/artifactory/scala-pr-validation-snapshots org.scala-lang:scala-compiler:2.13.2-bin-7cb1bd9-SNAPSHOT` scala.tools.nsc.MainGenericRunner -usejavacp
Welcome to Scala 2.13.2-20200325-024252-7cb1bd9 (Java HotSpot(TM) 64-Bit GraalVM EE 19.2.1, Java 1.8.0_231).
Type in expressions for evaluation. Or try :help.
Exception in thread "main" java.lang.NoClassDefFoundError: org/jline/reader/Completer
	at scala.tools.nsc.interpreter.shell.ILoop.defaultIn$lzycompute(ILoop.scala:67)
        ...
```

The dependency for JNA is kind of optional. The REPL works fine without JNA on macOS after adding `org.jline:jline:3.14.0` to `cs fetch`, I guess JLine uses a ["child processes whenever the terminal is accessed"](https://github.com/jline/jline3#jansi-vs-jna), not sure if that comes with its own set of problems? On windows we get "creating a dumb terminal". That will likely break tools that start the repl on windows (including sbt? I didn't test).

SethTisue commented:
> I think this is too breaking, at least for JLine, but IMO also for JNA

Okay, I've changed it back to regular dependencies, as the conservative choice, in the interest of getting this merged. (Eugene, if you think one or both dependencies should be optional, you could PR that separately and argue for it.)

We could consider it (we already consider it!) future work to move the whole REPL front end into its own JAR, and then put the JLine and JNA dependencies on that (instead of scala-compiler.jar).

eed3si9n commented:
> Eugene, if you think one or both dependencies should be optional, you could PR that separately and argue for it.

If we are ok with I'm fine with that too. For the tooling maintainers, it would be good to document the full dependency JARs (both for 2.13.2 and 2.13.1) somewhere before we forget all this.

som-snytt commented:
Forget what? 😺

SethTisue commented:
> For the tooling maintainers, it would be good to document the full dependency JARs (both for 2.13.2 and 2.13.1) somewhere before we forget all this

To the PR description above, I have now added:

> `scala-compiler.jar` now depends on `net.java.dev.jna#jna` (instead of `org.fusesource.jansi#jansi`, as formerly).

I'm not sure what else you might have in mind that we should document?

bishabosha commented:
When i try to use `:paste` with the following
```scala
abstract class D { val d: Any; type T }

class DDep[D2 <: (D { type T = d.type }) { val d: Int }] {
  def test_==(d: D2)(t: d.d.type): Boolean = d.d == t
}
```

it formats pretty strangely:
```scala
scala> :paste
// Entering paste mode (ctrl-D to finish)

abstract class D { val d: Any; type T }
     | 
     | class DDep[D2 <: (D { type T = d.type }) { val d: Int }] {
     |   def test_==(d: D2)(t: d.d.type): Boolean = d.d == t
     | }
```

eed3si9n commented:
> I'm not sure what else you might have in mind that we should document?

I was thinking somewhere more permanent like readme or scala-lang site, describing why/when JNA is needed.

som-snytt commented:
@bishabosha there was a "to do" for migrating the old line continuation code, but I can't find a ticket. Probably it's better to create an issue than comment here. (Some tickets are on scala-dev.)

SethTisue commented:
@bishabosha thanks for the report; opened https://github.com/scala/scala-dev/issues/703 and marked it "blocker"

@som-snytt you're probably thinking of https://github.com/scala/scala-dev/issues/689 ("Excise old continuation handling" — I'd assumed that was just cleanup, but now we have a case where the old code is actually interfering)

SethTisue commented:
> I was thinking somewhere more permanent like readme or scala-lang site, describing why/when JNA is needed.

@eed3si9n suggestion recorded at https://github.com/scala/scala-dev/issues/685

SethTisue commented:
#8848 sands down some rough spots in the behavior

twome commented:
This puts a file named `.scala_history_jline3` in the root of my home folder. I have no idea what "JLine3" is and no-one should ever need to websearch a mysterious file that appears in, again, *the root directory of their home folder* to figure out what it's for.  Can you please respect your users and at least put history/configuration/cache files inside a single folder named `.scala`?

SethTisue commented:
@twome please open an issue at https://github.com/scala/bug

twome commented:
@SethTisue I opened an issue but this isn't reporting a bug, this is reporting your software doing something shitty to my computer without my permission. If you're creating junk on people's computers it's *your* responsibility to remember and stop that.

som-snytt commented:
@twome I added my rant to your rant. I'd suggest future ranting on your ticket instead of graffiti on old PRs.
